### PR TITLE
Adding Virtual Network Usage cmdlet

### PR DESCRIPTION
**Pull request by EvgenyAgafonchikov**
## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

File: src/ResourceManager/Network/Commands.Network/VirtualNetwork/VirtualNetworkListUsageMethod.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        [Parameter(
+            Mandatory = true,
+            ValueFromPipelineByPropertyName = true)]
+        [ValidateNotNullOrEmpty]
+        public string ResourceGroupName { get; set; }
+        [Parameter(
+            Mandatory = true,
+            ValueFromPipelineByPropertyName = true)]
+        [ValidateNotNullOrEmpty]
+        public string Name { get; set; }
```
Comment (shahabhijeet) [inline src/ResourceManager/Network/Commands.Network/VirtualNetwork/VirtualNetworkListUsageMethod.cs]: @EvgenyAgafonchikov what does Name parameter mean here?

File: src/ResourceManager/Network/Commands.Network/help/Add-AzureRmRouteFilterRuleConfig.md [hunk 2]
```diff
@@ -35,6 +35,8 @@ The Add-AzureRmRouteFilterRuleConfig cmdlet adds a route filter rule to an Azure
 
 
 
+
```
Comment (shahabhijeet) [inline src/ResourceManager/Network/Commands.Network/help/Add-AzureRmRouteFilterRuleConfig.md]: @EvgenyAgafonchikov remove extra line breaks

File: src/ResourceManager/Network/Commands.Network/help/Get-AzureRmVirtualNetworkUsageList.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+## DESCRIPTION
+The **Get-AzureRmVirtualNetworkUsageList** cmdlet gets per subnet usage for the specified virtual network.
+
+## EXAMPLES
+
+### Example 1
+```
+PS C:\> Get-AzureRmVirtualNetworkUsageList -ResourceGroupName test -Name usagetest
+
+Get-AzureRmVirtualNetworkUsageList -ResourceGroupName test -Name usagetest
```
Comment (shahabhijeet) [inline src/ResourceManager/Network/Commands.Network/help/Get-AzureRmVirtualNetworkUsageList.md]: @EvgenyAgafonchikov can you provide an example of how to use this particular cmdlet when a list if returned
Give a full example something along the line of
$vNetUsageList = Get-AzureRmVirtualNetworkUsageList -ResourceGroupName test -Name usagetest
Because the name has the noun list, but all you are doing is returning 1 PSVirtualNetworkUsage rather the expectation is it should return IEnumerable<PSVirtualNetworkUsage>

File: src/ResourceManager/Network/Commands.Network/help/Add-AzureRmRouteFilterRuleConfig.md (hunk 2 above)
Comment (EvgenyAgafonchikov) [inline src/ResourceManager/Network/Commands.Network/help/Add-AzureRmRouteFilterRuleConfig.md]: @shahabhijeet, *.md files line breaks are added during the process of help autogeneration. Could you please clarify, should I manually update autogenerated files or keep them as is in this case?

File: src/ResourceManager/Network/Commands.Network/VirtualNetwork/VirtualNetworkListUsageMethod.cs (hunk 1 above)
Comment (EvgenyAgafonchikov) [inline src/ResourceManager/Network/Commands.Network/VirtualNetwork/VirtualNetworkListUsageMethod.cs]: `Name` is name of the Virtual Network to perform operation on. Added help messages to avoid confusion.

File: src/ResourceManager/Network/Commands.Network/help/Get-AzureRmVirtualNetworkUsageList.md (hunk 3 above)
Comment (EvgenyAgafonchikov) [inline src/ResourceManager/Network/Commands.Network/help/Get-AzureRmVirtualNetworkUsageList.md]: Updated cmdlet to return `List<PSVirtualNetworkUsage>` explicitly.
The example output sample contains list of two `PSVirtualNetworkUsage` items now

File: src/ResourceManager/Network/Commands.Network/Models/PSVirtualNetworkUsage.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+﻿namespace Microsoft.Azure.Commands.Network.Models
+{
+    public partial class PSVirtualNetworkUsage
+    {
+        public string Id;
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/Models/PSVirtualNetworkUsage.cs]: @EvgenyAgafonchikov nit: for each of these properties, can we add the `get` and `set` methods to be consistent with the other model classes?

```cs
public string Id { get; set; }
```

File: src/ResourceManager/Network/Commands.Network/VirtualNetwork/VirtualNetworkListUsageMethod.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            base.Execute();
+
+            var vVirtualNetworkList = this.NetworkClient.NetworkManagementClient.VirtualNetworks.ListUsage(ResourceGroupName, Name);
+            var vnetUsageList = new List<PSVirtualNetworkUsage>();
+            foreach (var vVirtualNetwork in vVirtualNetworkList)
+            {
+                var vVirtualNetworkModel = Mapper.Map<CNM.PSVirtualNetworkUsage>(vVirtualNetwork);
+                vnetUsageList.Add(vVirtualNetworkModel);
+            }
+            WriteObject(vnetUsageList);
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/VirtualNetwork/VirtualNetworkListUsageMethod.cs]: @EvgenyAgafonchikov this should be `WriteObject(vnetUsageList, true)` (which sets the `enumerateCollection` parameter to true) so that the list you have constructed writes each object to the output stream rather than writes the array as a whole object to the stream

File: src/ResourceManager/Network/Commands.Network/help/Add-AzureRmRouteFilterRuleConfig.md (hunk 2 above)
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/help/Add-AzureRmRouteFilterRuleConfig.md]: @EvgenyAgafonchikov `platyPS` will sometimes add these blank lines (I haven't determined why yet), so you can go ahead and revert the changes that add unnecessary lines that are autogenerated.

File: src/ResourceManager/Network/Commands.Network/help/Get-AzureRmApplicationGatewayBackendHealth.md [hunk 6]
```diff
@@ -37,6 +37,8 @@ The Get-AzureRmApplicationGatewayBackendHealth cmdlet gets application gateway b
 
 
 
+
+
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/help/Get-AzureRmApplicationGatewayBackendHealth.md]: @EvgenyAgafonchikov same comment about removing the extra blank lines in this file

File: src/ResourceManager/Network/Commands.Network/help/Test-AzureRmNetworkWatcherConnectivity.md [hunk 7]
```diff
@@ -38,6 +38,8 @@ The Test-AzureRmNetworkWatcherConnectivity cmdlet returns connectivity informati
 
 
 
+
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/help/Test-AzureRmNetworkWatcherConnectivity.md]: @EvgenyAgafonchikov same comment about removing the blank lines in this file

**Comment by EvgenyAgafonchikov** [comment]
@cormacpayne , I've processed recent comments, please have a look once CI pass

**Comment by cormacpayne** [comment]
On-demand: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1707/

**Comment by shahabhijeet** [comment]
@EvgenyAgafonchikov looks good, should be good to go, once all @cormacpayne feedback is addressed.
